src/NHibernate/Cache/FilterKey.cs

```diff
@@ -8,10 +10,16 @@
 namespace NHibernate.Cache
 {
 	[Serializable]
-	public class FilterKey
+	public class FilterKey : IDeserializationCallback
 	{
 		private readonly string _filterName;
-		private readonly Dictionary<string, TypedValue> _filterParameters = new Dictionary<string, TypedValue>();
+
+		// Sets and dictionaries are populated last during deserialization, causing them to be potentially empty
+		// during the deserialization callback. This causes them to be unreliable when used in hashcode computations.
+		// So better serialize them as other structures and use those other structures in hashcode computation.
+		[NonSerialized]
+		private Dictionary<string, TypedValue> _filterParameters = new Dictionary<string, TypedValue>();
```

Comment:
Since `_filterParameters` field is not used outside of the class it would be easier to _just_ store it as a serializable collection (eg array).

Reply:
This will have a performance impact on `Equals`, switching an O(n) algorithm for an an O(n²) one. That said, `n` should be always quite low in most cases, so it may not be noticeable at all.